File: src/main/vrml/net/coderextreme/x3d_code/www_web3d_org/x3d/content/examples/Basic/development/ProxyShapeExample.js
let browser = X3D.getBrowser();
let X3D0 = {};
X3D0.profile = "Immersive";
X3D0.version = "3.0";
let WorldInfo2 = browser.currentScene.createNode("WorldInfo");
WorldInfo2.title = "ProxyShapeExample.x3d";
browser.currentScene.children = new MFNode();

browser.currentScene.children[0] = WorldInfo2;

let Collision3 = browser.currentScene.createNode("Collision");
Collision3.DEF = "MyCollision";
//can experiment with order of proxy and children nodes without error in XML encoding, but reordering difficulties can occur in other programming language bindings.
let Shape4 = browser.currentScene.createNode("Shape");
let Sphere5 = browser.currentScene.createNode("Sphere");
Shape4.geometry = Sphere5;

let Appearance6 = browser.currentScene.createNode("Appearance");
Appearance6.DEF = "RedAppearance";
let Material7 = browser.currentScene.createNode("Material");
Material7.diffuseColor = new SFColor(new float[1,0,0]);
Appearance6.material = Material7;

Shape4.appearance = Appearance6;

Collision3.proxy = Shape4;

let Shape8 = browser.currentScene.createNode("Shape");
let Cone9 = browser.currentScene.createNode("Cone");
Shape8.geometry = Cone9;

let Appearance10 = browser.currentScene.createNode("Appearance");
Appearance10.USE = "RedAppearance";
Shape8.appearance = Appearance10;

Collision3.proxy = Shape8;

let Shape11 = browser.currentScene.createNode("Shape");
let Box12 = browser.currentScene.createNode("Box");
Box12.size = new SFVec3f(new float[16,16,16]);
Shape11.geometry = Box12;

Collision3.proxy = Shape11;

browser.currentScene.children[1] = Collision3;

let Script13 = browser.currentScene.createNode("Script");
Script13.DEF = "MyNodeProducer";
Script13.directOutput = True;
let field14 = browser.currentScene.createNode("field");
field14.name = "outputNode";
field14.accessType = "outputOnly";
field14.type = "MFNode";
Script13.field = new MFNode();

Script13.field[0] = field14;


Script13.setSourceCode(`ecmascript:\n"+
"\n"+
"function initialize ()\n"+
"{\n"+
"	nodesString ='Group { children [ Shape {' +\n"+
"	'geometry Text {' + \n"+
"		 'string [ \"proxy is\" \"any-order child\" \"of Collision\" ]' +\n"+
"		 'fontStyle FontStyle { justify [ \"MIDDLE\" \"MIDDLE\" ] } }' +\n"+
"	'appearance Appearance {' +\n"+
"		 'material Material { diffuseColor 0.1  0.4 0.8 }' + \n"+
"	'} } ] }';\n"+
"	Browser.print (nodesString);\n"+
"	outputNode = Browser.createVrmlFromString (nodesString);\n"+
"}`)
browser.currentScene.children[2] = Script13;

let ROUTE15 = browser.currentScene.createNode("ROUTE");
ROUTE15.fromField = "outputNode";
ROUTE15.fromNode = "MyNodeProducer";
ROUTE15.toField = "children";
ROUTE15.toNode = "MyCollision";
browser.currentScene.children[3] = ROUTE15;

File: src/main/python/net/x3djsonld/data/c/x3d-code/www.web3d.org/x3d/content/examples/Basic/development/ProxyShapeExample.py
###############################################
#
# Now available: developmental python x3d.py package on PyPi for import.
#   This approach greatly simplifies Python X3D deployment and use.
#   https://pypi.org/project/x3d
#
# Installation:
#       pip install x3d
# or
#       python -m pip install x3d
#
# Developer options for loading x3d package:
#
#    from x3d import *  # preferred approach, terser source that avoids x3d.* class prefixes
#
# or
#    import x3d         # traditional way to subclass x3d package, all classes require x3d.* prefix
#                       # but python source is very verbose, for example x3d.Material x3d.Shape etc.
#                       # X3dToPython.xslt stylesheet insertPackagePrefix=true supports this option.

from x3d import *

###############################################

newModel=X3D(profile='Immersive',version='3.0',
  head=head(
    children=[
    meta(content='ProxyShapeExample.x3d',name='title'),
    meta(content='Don Brutzman',name='creator'),
    meta(content='14 January 2000',name='created'),
    meta(content='20 October 2019',name='modified'),
    meta(content='Collision node, child test case: can reorder children nodes as Shape Proxy Shape or other variations without problems in XML encoding, but reordering difficulties can occur in other programming language bindings.',name='description'),
    meta(content='https://www.web3d.org/technicalinfo/specifications/vrml97/part1/nodesRef.html#Collision',name='reference'),
    meta(content='https://www.web3d.org/x3d/content/examples/Basic/development/ProxyShapeExample.x3d',name='identifier'),
    meta(content='X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit',name='generator'),
    meta(content='../license.html',name='license')]),
  Scene=Scene(
    children=[
    WorldInfo(title='ProxyShapeExample.x3d'),
    Collision(DEF='MyCollision',
      #  can experiment with order of proxy and children nodes without error in XML encoding, but reordering difficulties can occur in other programming language bindings. 
      proxy=Shape(
        geometry=Box(size=(16,16,16))),
      children=[
      Shape(
        geometry=Sphere(),
        appearance=Appearance(DEF='RedAppearance',
          material=Material(diffuseColor=(1,0,0)))),
      Shape(
        geometry=Cone(),
        appearance=Appearance(USE='RedAppearance'))]),
    Script(DEF='MyNodeProducer',directOutput=True,
      field=[
      field(accessType='outputOnly',name='outputNode',type='MFNode')]),
    ROUTE(fromField='outputNode',fromNode='MyNodeProducer',toField='children',toNode='MyCollision')])
) # X3D model complete

###############################################
# Self-test diagnostics
###############################################

if        metaDiagnostics(newModel): # built-in utility method in X3D class
    print(metaDiagnostics(newModel))
print('check  newModel.XML() serialization...')
newModelXML = newModel.XML() # test export method XML() for exceptions
# print(newModelXML) # debug

print ("python x3d.py load successful for ProxyShapeExample.py")
